src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -835,28 +840,39 @@ private ImmutableArray<CompilationEvent> DequeueGeneratedCompilationEvents()
 
         private async Task<AnalyzerDriver> GetAnalyzerDriverAsync(CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
-
-            // Get instance of analyzer driver from the driver pool.
-            AnalyzerDriver driver = _driverPool.Allocate();
-
             try
             {
-                // Start the initialization task, if required.
-                if (driver.WhenInitializedTask == null)
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Get instance of analyzer driver from the driver pool.
+                AnalyzerDriver driver = _driverPool.Allocate();
+
+                bool success = false;
+                try
                 {
-                    driver.Initialize(_compilation, _analysisOptions, _compilationData, categorizeDiagnostics: true, cancellationToken: cancellationToken);
-                }
+                    // Start the initialization task, if required.
+                    if (driver.WhenInitializedTask == null)
+                    {
+                        driver.Initialize(_compilation, _analysisOptions, _compilationData, categorizeDiagnostics: true, cancellationToken: cancellationToken);
+                    }
 
-                // Wait for driver initialization to complete: this executes the Initialize and CompilationStartActions to compute all registered actions per-analyzer.
-                await driver.WhenInitializedTask.ConfigureAwait(false);
+                    // Wait for driver initialization to complete: this executes the Initialize and CompilationStartActions to compute all registered actions per-analyzer.
+                    await driver.WhenInitializedTask.ConfigureAwait(false);
 
-                return driver;
+                    success = true;
+                    return driver;
+                }
+                finally
+                {
+                    if (!success)
+                    {
+                        FreeDriver(driver);
+                    }
+                }
             }
-            catch (OperationCanceledException)
+            catch (Exception e) when (FatalError.ReportUnlessCanceled(e))
```

Comment:
❔  What's the intended stack here? Currently if the exception is thrown before the `finally` block, it will capture the stack later when the `finally` block completes.

Reply:
I'm not sure I understand the question. There are two changes here: 
1) retrhow is removed, so that any OperationCancelledException preserves its original stack trace
2) The whole body is wrapped in a filter, which triggers fail-fast if any exception, other than OperationCanceledException, is thrown and not handled within the code that is called form the method body.

Reply:
There is a difference between the way this method is implemented, and the way GenerateCompilationEventsAndPopulateEventsCacheAsync is getting handled by its caller. The inconsistency makes it more difficult for future maintainers to identify the pattern to apply to new cases.

```csharp
// Case 1:
try
{
  try
  {
    // Begin inlined callee
    try
    {
      innerOperation();
    }
    catch (Exception e) when (FatalError.ReportUnlessCancelled(e))
    {
      throw ExceptionUtilities.Unreachable;
    }
    // End inlined callee
  }
  finally
  {
    // cleanup from failure
  }
}
catch (Exception e) when (FatalError.ReportUnlessCanceled(e))
{
  throw ExceptionUtilities.Unreachable;
}

// Case 1:
try
{
  try
  {
    // Begin inlined callee
    innerOperation();
    // End inlined callee
  }
  finally
  {
    // cleanup from failure
  }
}
catch (Exception e) when (FatalError.ReportUnlessCanceled(e))
{
  throw ExceptionUtilities.Unreachable;
}
```

> retrhow is removed, so that any OperationCancelledException preserves its original stack trace

😕 I thought the whole point of `throw;` is it wouldn't interfere with the stack trace.

Reply:
The pattern for the fail-fast filter is to wrap a body of an async method with try-catch-when. That's because the compiler emits try-catch around every async method body, so that it can store the caught exception into the Task. That however unwinds the stack, which is why we need to insert the exception filter that captures the memory dump before that happens. 

The same applies to async lambdas, which is why we have another fail-fast filter in the first case above. 

>  I thought the whole point of throw; is it wouldn't interfere with the stack trace.

Rethrow always unwinds the stack. It preserves the _stack trace_, but that's not useful when you're looking for values of local variables in a memory dump.